### PR TITLE
Further improvements

### LINKED ISSUE #278: /usr/lib/postgresql/*/bin not in PATH
Hi,
I was just trying to call `/usr/lib/postgresql/13/bin/ogr_fdw_info` but it looks that `/usr/lib/postgresql/13/bin` is not included in `$PATH`.

### PR DESCRIPTION
**Pull request by NyakudyaA**
* Added the base build directory to the main repo
* Removed some duplications elsewhere
* Revert commit https://github.com/kartoza/docker-postgis/commit/e6bc0fb17dc80fa657fd58d35fdfa056bacf4ff5
* Fixes https://github.com/kartoza/docker-postgis/issues/278
* Upgrade container to use gdal3. Debian buster still install gdal 2.4 and PostGIS 13 is best served with gdal3
* Upgrade minor release to 13.1

PS: We can integrate GitHub workflows to build the base image

### CODE REVIEW

**Comment by NyakudyaA** [comment]
@lucernae I am not sure why the tests are failing but I tested on my local and all is working well

**Comment by NyakudyaA** [comment]
ping @lucernae

**Comment by lucernae** [comment]
I think I already fixed this before. Basically it’s just because the main image built are not built yet. Let me double check

**Comment by lucernae** [comment]
Ok, it’s a different error. The base image failed to build, that’s why kartoza/postgis:manual-build is not ready yet.
There seems to be an issue in step:

```

E: Sub-process /usr/bin/dpkg returned an error code (1)
2608The command '/bin/sh -c set -eux     && export DEBIAN_FRONTEND=noninteractive     &&  apt-get update     && apt-get -y --no-install-recommends install postgresql-client-${POSTGRES_MAJOR_VERSION}         postgresql-common postgresql-${POSTGRES_MAJOR_VERSION}         postgresql-${POSTGRES_MAJOR_VERSION}-postgis-${POSTGIS_MAJOR}         netcat postgresql-${POSTGRES_MAJOR_VERSION}-ogr-fdw         postgresql-${POSTGRES_MAJOR_VERSION}-postgis-${POSTGIS_MAJOR}-scripts         postgresql-plpython3-${POSTGRES_MAJOR_VERSION} postgresql-${POSTGRES_MAJOR_VERSION}-pgrouting         postgresql-server-dev-${POSTGRES_MAJOR_VERSION} postgresql-${POSTGRES_MAJOR_VERSION}-cron' returned a non-zero code: 100

```

We should probably do `set -eux` in the build script so that it fails immediately when the build failed. Can’t tell much what happens in that commend. Probably missing env?

**Comment by lucernae** [comment]
<strike>I got this from stderr:

```
W: GPG error: https://deb.meteo.guru/debian buster InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6BB3BABB751F80D7
E: The repository 'https://deb.meteo.guru/debian buster InRelease' is not signed.
```

This apt update right after using the base image. so we need to fix the base image.</strike>

Edit: sorry, it still picks up old branch caches. had to do a full build. Disregard above comments.

**Comment by NyakudyaA** [comment]
> I got this from stderr:
> 
> ```
> W: GPG error: https://deb.meteo.guru/debian buster InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6BB3BABB751F80D7
> E: The repository 'https://deb.meteo.guru/debian buster InRelease' is not signed.
> ```
> 
> This apt update right after using the base image. so we need to fix the base image.
> 
> Edit: sorry, it still picks up old branch caches. had to do a full build. Disregard above comments.

Roger this

**Comment by lucernae** [comment]
@NyakudyaA Try pulling from https://github.com/kartoza/docker-postgis/tree/fix-ci
I've fixed some issues with the build, so now it builds using multi-stage.
Deployment at local can be done via `docker-compose build` with argument supplied from `.env` file.
Build test in CI was done via Github Action Workflows (like we wanted to do in the beginning of the PR), then we can replace Travis.

I added some custom build args too, so we probably need to update the docs. I will explain it in PR review after you do the pull from fix-ci to your branch.

<strike>I was not able to repair all the scenario tests, it seems the others are really broken.</strike>

Fixed now! There are some typos in the conf generator. So the tests indeed broke.

File: .example.env [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# Set GENERATE_ALL_LOCALE to empty value or 0 to build just default LOCALE: en_US.UTF-8
+GENERATE_ALL_LOCALE=1
+# Set the language if you need to specify LANG locale at build time
+LANG=en_US.UTF-8
+# locale filter to include in the locale generator
+LANGS="en_US.UTF-8,id_ID.UTF-8"
+
+POSTGRES_MAJOR_VERSION=13
+POSTGIS_MAJOR_VERSION=3
+POSTGIS_MINOR_RELEASE=1
```
Comment (lucernae) [inline .example.env:18]: Build arguments (and runtime docker-compose arguments) must be declared here.

Local user can do custom local build by copying this template into `.env` and fill in their values.
When we have changes to this file, user can diff their `.env` to this file for reference.

File: .example.env [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+## For build arguments
+DISTRO=debian
+IMAGE_VERSION=bullseye
+IMAGE_VARIANT=slim
+# Set GENERATE_ALL_LOCALE to empty value or 0 to build just default LOCALE: en_US.UTF-8
+GENERATE_ALL_LOCALE=1
+# Set the language if you need to specify LANG locale at build time
+LANG=en_US.UTF-8
+# locale filter to include in the locale generator
+LANGS="en_US.UTF-8,id_ID.UTF-8"
```
Comment (lucernae) [inline .example.env:14]: I added this build args to only build default locale.
Useful for creating slim base images to do the scenario tests (less time to build).
The reason why we use both en_US and id_ID is because it's used in the tests. So this is the default.

Both `LANG` and `LANGS` only used if `GENERATE_ALL_LOCALE=0` or empty (not generating all locale).

File: .example.env (hunk 1 above)
Comment (lucernae) [inline .example.env:18]: I extracted these:
```
POSTGRES_MAJOR_VERSION=13
POSTGIS_MAJOR_VERSION=3
POSTGIS_MINOR_RELEASE=1
```

From the Dockerfile recipe. Also included `POSTGIS_MAJOR_VERSION` to make it easier to understand.

File: .github/workflows/deploy-image.yaml [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+          password: ${{ secrets.DOCKERHUB_PASSWORD }}
+
+      - name: Build base image
+        id: docker_build_base
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: Dockerfile
+          push: true
+          tags: ${{ secrets.DOCKERHUB_REPO }}/postgis:base-${{ matrix.distro }}-${{ matrix.imageVersion }}-${{ matrix.imageVariant }}
```
Comment (lucernae) [inline .github/workflows/deploy-image.yaml:74]: RFC to @NyakudyaA .

I propose we switch the image tag from `debian-bullseye-slim-base` into `base-debian-bullseye-slim`, due to the public convention that image variant is always the last, and `-` separates context from generic to specific (`base` being more generic usage).

Ideally there should be also the latest base tag which only spells `base`, like public convention.

File: .github/workflows/deploy-image.yaml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+      - name: Build prod image
+        id: docker_build_prod
+        uses: docker/build-push-action@v2
+        with:
+          context: .
+          file: Dockerfile
+          push: true
+          load: true
+          tags: ${{ secrets.DOCKERHUB_REPO }}/postgis:${{ matrix.postgresMajorVersion }}-${{ matrix.postgisMajorVersion }}.${{ matrix.postgisMinorRelease }}
```
Comment (lucernae) [inline .github/workflows/deploy-image.yaml:95]: RFC to @NyakudyaA .

I propose we should also change the tag from `13.1` into `13-3.1` where the first part before dash is the postgres major version, and the rest is postgis full version to reduce ambiguity. 

Ideally we should also define the orchestration version somewhere in the back, like `13-3.1-21.01` where `21.01` is a shorthand for 2021 January. I prefer using `21.01` for orchestration tag (to version the Dockerfile recipe or build scripts) instead of semver because we don't yet have any clear convention to define a 'feature'.

The specific tag can be used for production usage (to really pin down the image), like `13-3.1-21.01.19`. But we should also track the common uses tag like `13` or `13.1` where user can easily type it for testing.

In practice, the common tag will always move. Today it can be that `13` is the same image as `13.1`, `13-3.1`, or `13-3.1-21.01` (in the order of specificity). Next month it could be that it points to `13-3.1-21-02`.

The one who will keep the tags to be up to date is the docker-meta actions (not yet being put here).

File: Dockerfile [hunk 5]
```diff
@@ -1,19 +1,85 @@
-#--------- Generic stuff all our Dockerfiles should start with so we get caching ------------
+##############################################################################
```
Comment (lucernae) [inline Dockerfile:1]: RFC to @NyakudyaA .

We are going to try to leverage Dockerfile Multi-stage build. I don't see any reasoning not too. It's more convenient, easier to maintain, and also cacheable. The only drawback that I regret is they can't be separate with some kind of `include` mechanism.

TL;DR; on how to leverage multi stage build:

We combine previous 3 Dockerfile (base, prod, and testing) into one but with 3 stages (base, prod, and testing).

Append extra parameters `--target` into the docker build (or docker-compose build) to specify which target you want to build.

Since we can refer the image inside the recipe (for example `postgis-base` is the name of the base image stage), caching happens without us having to push the image into registry. For example the  prod image is built using `FROM postgis-base AS postgis-prod`. We can do build like this:

```
# Build base image first (like we usually do)
docker build --target postgis-base .
# Build prod image, but it will start by using the cache from the postgis-base stage (no need to rebuild base)
docker build --target postgis-prod .
```

Or like this:
```
# Build prod image directly
docker build --target postgis-prod .
# Then build the base image as a tagged image, because we want to push the base over to registry
docker build -t kartoza/postgis:base --target postgis-base .
# In the above command, docker will use build cache because it already build postgis-base stage, when it builds postgis-prod
```

Basically we can keep the same usage workflow as usual, but with caching improvement in local and cloud build. The only thing we need to change is the build scripts.

File: README.md [hunk 6]
```diff
@@ -59,8 +59,7 @@ docker pull kartoza/postgis:image_version
 
 ## Building the image
 
-To build the image yourself without apt-cacher (also consumes more bandwidth
-since deb packages need to be fetched each time you build) do:
+To build the image yourself do:
```
Comment (lucernae) [inline README.md:62]: We need to update the docs, once this improvements are approved/finalized

File: base_build/scripts/locale-filter.sh [hunk 7]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+## Filter list of locales from a given filter args
+## Parse into array
+LANG_ARR=(${LANGS//,/ })
+echo "" > /etc/locale.gen
+for i in "${LANG_ARR[@]}"; do
+  cat /etc/all.locale.gen | grep "$i" >> /etc/locale.gen
+done
```
Comment (lucernae) [inline base_build/scripts/locale-filter.sh:9]: This is a helper script to filter desired locale in the build stage of postgis-base.
The scripts is unused if we build all locale

File: build-base.sh [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+# Building an debian base image
+
+if [[ ! -f .env ]]; then
+    echo "Default build arguments don't exists. Creating one from default value."
+    cp .example.env .env
+fi
+
+docker-compose -f docker-compose.build.yml build postgis-base
+
```
Comment (lucernae) [inline build-base.sh:10]: I moved this from `base_build` and adapt it to use docker-compose, multi-stage build and the `.env` file for build args

File: build-test.sh [hunk 9]
```diff
@@ -1,5 +1,9 @@
 #!/usr/bin/env bash
+# For scenario testing purposes
 
-./build.sh
```
Comment (lucernae) [inline build-test.sh:3]: No need to specify base build. Because docker will understand the requirements from the stage declaration

File: build-test.sh [hunk 10]
```diff
@@ -1,5 +1,9 @@
...
+if [[ ! -f .env ]]; then
+    echo "Default build arguments don't exists. Creating one from default value."
+    cp .example.env .env
+fi
 
-docker build -t kartoza/postgis:manual-build -f Dockerfile.test .
+docker-compose -f docker-compose.build.yml build postgis-test
```
Comment (lucernae) [inline build-test.sh:9]: Simplify build command. Now the tags are declared in the docker-compose file. The stage are also declared in the docker-compose file. We only need to specify which service to build.

File: build.sh [hunk 11]
```diff
@@ -1,3 +1,8 @@
 #!/usr/bin/env bash
-docker build -t kartoza/postgis:manual-build .
-docker build -t kartoza/postgis:13.0 .
```
Comment (lucernae) [inline build.sh:3]: Doesn't need to declare it. The dependencies are declared from the multi-stage build and declared in docker-compose file.

File: docker-compose.build.yml [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+      context: .
+      # Use yaml anchor for reusability
+      args: &build-args-anchor
+        DISTRO: ${DISTRO}
+        IMAGE_VERSION: ${IMAGE_VERSION}
+        IMAGE_VARIANT: ${IMAGE_VARIANT}
+        GENERATE_ALL_LOCALE: ${GENERATE_ALL_LOCALE}
+        POSTGRES_MAJOR_VERSION: ${POSTGRES_MAJOR_VERSION}
+        POSTGIS_MAJOR_VERSION: ${POSTGIS_MAJOR_VERSION}
+        POSTGIS_MINOR_RELEASE: ${POSTGIS_MINOR_RELEASE}
```
Comment (lucernae) [inline docker-compose.build.yml:16]: Using YAML anchor because all the build args used in all stages must be the same.

File: docker-compose.build.yml [hunk 13]
```diff
@@ -0,0 +1,33 @@
+# Used solely for docker-compose build
```
Comment (lucernae) [inline docker-compose.build.yml:1]: This file is specific for building only. No runtime spec declared.

File: scenario_tests/collations/test.sh [hunk 14]
```diff
@@ -8,17 +8,23 @@ source ../test-env.sh
 # Run service
 docker-compose up -d
 
+if [[ -n "${PRINT_TEST_LOGS}" ]]; then
+  docker-compose logs -f &
+fi
+
```
Comment (lucernae) [inline scenario_tests/collations/test.sh:14]: All the scenarios must have this so we can see the log in GH Action log as the tests runs asynchronously.

File: scenario_tests/collations/test.sh [hunk 15]
```diff
@@ -8,17 +8,23 @@ source ../test-env.sh
...
 sleep 30
 
 services=("pg" "pg-new")
 
 for service in "${services[@]}"; do
 
   # Execute tests
-  until docker-compose exec $service pg_isready; do
+  until docker-compose exec -T $service pg_isready; do
```
Comment (lucernae) [inline scenario_tests/collations/test.sh:22]: All `docker-compose exec` must have `-T` flag so we can see the output in Github Action (it means no pseudo TTY allocation). But this is also harmless in local tests

File: scenario_tests/logical_replication/docker-compose.yml [hunk 16]
```diff
@@ -47,8 +47,7 @@ services:
       REPLICATION_PASS: 'replicator'
       REPLICATION: 'true'
     depends_on:
-        pg-publisher:
-          condition: service_healthy
+        - pg-publisher
```
Comment (lucernae) [inline scenario_tests/logical_replication/docker-compose.yml:50]: I had to explicitly do this. Because we don't want to wait for `pg-subscriber` to be up to show the logs. If something were happen to the service (causing `pg-publisher` not healthy), it will just wait block unless we declare it like this. So this is necessary for testing purposes.

File: scripts/setup-conf.sh [hunk 17]
```diff
@@ -39,12 +38,14 @@ timezone='${TIMEZONE}'
...
-
+echo "include 'postgis.conf'" >> $CONF
 
+# Create a config for logical replication
 if [[  "${REPLICATION}" =~ [Tt][Rr][Uu][Ee] && "$WAL_LEVEL" == 'logical' ]]; then
-cat >> "$CONF" <<EOF
+  if [[  -f ${ROOT_CONF}/logical_replication.conf ]];then
+    rm $CONF/logical_replication.conf
+  fi
+cat >> ${ROOT_CONF}/logical_replication.conf <<EOF
```
Comment (lucernae) [inline scripts/setup-conf.sh:48]: I fixed some typo. It was previously `streaming_replication.conf` so the test fails in `logical_replication` scenario

File: .github/workflows/build-latest.yaml [hunk 18]
```diff
@@ -0,0 +1,197 @@
...
+#        required: true
+#        default: debian
+#      imageVersion:
+#        description: Base distro image version/release
+#        required: true
+#        default: bullseye
+#      imageVariant:
+#        description: Base image variant
+#        required: true
+#        default: slim
```
Comment (lucernae) [inline .github/workflows/build-latest.yaml:16]: This is commented for now, but can be useful if we have a use case to do an ad-hoc cloud build via GH Action.

For example, if the recipe can handle different base image. We can try to do a build from GH UI.

File: .github/workflows/build-latest.yaml [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+          path: /tmp/.buildx-cache
+          # Build inputs are:
+          # - Dockerfile recipe
+          # - docker-compose.build.yml
+          # - build args (.example.env)
+          # - base_build directory
+          # - scripts directory
+          key: buildx-${{ hashFiles('Dockerfile', 'docker-compose.build.yml', '.example.env', 'base_build', 'scripts') }}-${{ github.sha }}
+          restore-keys: |
+            buildx-${{ hashFiles('Dockerfile', 'docker-compose.build.yml', '.example.env', 'base_build', 'scripts') }}-
```
Comment (lucernae) [inline .github/workflows/build-latest.yaml:54]: HashFiles of the build inputs are used to determine if we are doing the same build or not.
If it's the same, just retrieve from cache.

File: .github/workflows/build-latest.yaml [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+          file: Dockerfile
+          push: false
+          load: true
+          tags: ${{ secrets.DOCKERHUB_REPO }}/postgis:base-${{ matrix.distro }}-${{ matrix.imageVersion }}-${{ matrix.imageVariant }}
+          build-args: |
+            DISTRO=${{ matrix.distro }}
+            IMAGE_VERSION=${{ matrix.imageVersion }}
+            IMAGE_VARIANT=${{ matrix.imageVariant }}
+            LANGS=en_US.UTF-8,id_ID.UTF-8
+            GENERATE_ALL_LOCALE=0
```
Comment (lucernae) [inline .github/workflows/build-latest.yaml:76]: For testing purposes, don't generate all locale.

**Review by lucernae — COMMENTED** [review]
Ok. Here's my proposed changes:

Some changes (RFCs) needed some comment back because it will affect our convention in the future.
Thumbs up if you like the RFC. Otherwise we can discuss and make approriate changes in this PR.

File: .github/workflows/deploy-image.yaml (hunk 4 above)
Comment (NyakudyaA) [reply .github/workflows/deploy-image.yaml:95]: I am ok with this. The convention was always that we always had one or two minor releases in each cycle. 

Something related to this versioning. Do we have a way to track if there is a new minor release so that we can automatically publish the corresponding image. Maybe we could investigate if we cannot do so for now

File: base_build/scripts/locale-filter.sh (hunk 7 above)
Comment (NyakudyaA) [reply base_build/scripts/locale-filter.sh:9]: Long overdue. Been a pending issue for people building locally a d Travis tests

File: scenario_tests/logical_replication/docker-compose.yml (hunk 16 above)
Comment (NyakudyaA) [reply scenario_tests/logical_replication/docker-compose.yml:50]: Maybe we could upgrade docker-compose from 2.1 to the latest version. The condition of service has been deprecated

File: .github/workflows/build-latest.yaml (hunk 18 above)
Comment (NyakudyaA) [reply .github/workflows/build-latest.yaml:16]: Not much experience with actions

File: .github/workflows/deploy-image.yaml (hunk 4 above)
Comment (lucernae) [reply .github/workflows/deploy-image.yaml:95]: @NyakudyaA , did you mean tracking new minor release from upstream? (like from Postgis source repo?). If this is the case, we can only do that via dependency bot.

Comment (NyakudyaA) [reply .github/workflows/deploy-image.yaml:95]: Yes, But we can do it in another PR

**Comment by lucernae** [comment]
@NyakudyaA , I can see that you agreed with the RFC.
Can we merge this so we can test action for build against all locale?

List of changes that can only be tested after this is merged in develop branch:
- Build against all locale
- Auto push to registry (merging this PR means we are trying to auto push `base-debian-bullseye-slim` and `13-3.1`
- Add short docker tag (like `base` and `13`)
- Do local build without push/login to docker registry on pull request (internal PR and external PR)
- Push test images on internal PR (PR from branches in this repo against develop)

Note that one of the workflow action above failing is because this is PR from external repo and we are using docker login inside. It failed because the Nyakudya:develop cannot access our secret (intentional for security). So, IMO, it's ok to ignore that for now. I'm going to fix that once the action is in develop.

**Comment by NyakudyaA** [comment]
> @NyakudyaA , I can see that you agreed with the RFC.
> Can we merge this so we can test action for build against all locale?
> 
> List of changes that can only be tested after this is merged in develop branch:
> 
>     * Build against all locale
> 
>     * Auto push to registry (merging this PR means we are trying to auto push `base-debian-bullseye-slim` and `13-3.1`
> 
>     * Add short docker tag (like `base` and `13`)
> 
>     * Do local build without push/login to docker registry on pull request (internal PR and external PR)
> 
>     * Push test images on internal PR (PR from branches in this repo against develop)
> 
> 
> Note that one of the workflow action above failing is because this is PR from external repo and we are using docker login inside. It failed because the Nyakudya:develop cannot access our secret (intentional for security). So, IMO, it's ok to ignore that for now. I'm going to fix that once the action is in develop.

LGTM @lucernae You can proceed to merge. So in future if want to merge from my branch it will always fail because of the security for secrets. Is it safe to assume that if all the other tests are passing then the image is fine regardless of failure/success to push to dockerhub

**Comment by lucernae** [comment]
> LGTM @lucernae You can proceed to merge. So in future if want to merge from my branch it will always fail because of the security for secrets. Is it safe to assume that if all the other tests are passing then the image is fine regardless of failure/success to push to dockerhub

We are going to do local build (inside github workflow) for external PR as it doesn't need docker login. (No need to setup secret for other user).
For internal PR, it's best that we push the PR image, so we can build the PR image. Something like kartoza/postgis:pr-123 then test it locally in our machine if we wish to. That means, checking PR from ourselves in our local machine doesn't need a local build. We just pull it from registry.

**Comment by NyakudyaA** [comment]
> > LGTM @lucernae You can proceed to merge. So in future if want to merge from my branch it will always fail because of the security for secrets. Is it safe to assume that if all the other tests are passing then the image is fine regardless of failure/success to push to dockerhub
> 
> We are going to do local build (inside github workflow) for external PR as it doesn't need docker login. (No need to setup secret for other user).
> For internal PR, it's best that we push the PR image, so we can build the PR image. Something like kartoza/postgis:pr-123 then test it locally in our machine if we wish to. That means, checking PR from ourselves in our local machine doesn't need a local build. We just pull it from registry.

Great @lucernae, when this is merged I think I will also adapt it to the GeoServer repo

**Comment by lucernae** [comment]
I will merge this now
